### ISSUE TITLE
Trigger sync immediately for shouldExecuteAnotherSyncHook

### ISSUE THREAD

**Issue by LukasKorba** [Electric-Coin-Company/zcash-swift-wallet-sdk#1345]
This hook already exists but does nothing compared to other cases like wipe and rewind. When the start of a sync is required during another sync in progress, the hook is created and CompactBlockProcessor rightfully processes the enum. There's no execution whatsoever though, see the code:

```swift
if let wipeContext = afterSyncHooksManager.shouldExecuteWipeHook() {
            try await doWipe(context: wipeContext)
            return false
        } else if let rewindContext = afterSyncHooksManager.shouldExecuteRewindHook() {
            try await doRewind(context: rewindContext)
            return false
        } else if afterSyncHooksManager.shouldExecuteAnotherSyncHook() {
            logger.debug("Starting new sync.")
            return true
        } else {
            return false
        }
```

### Goal
The sync should immediately start, otherwise it will happen automatically in 20-30s like normal sync. This delay is unwanted though, prolonging the UX of syncing, especially noticeable in cases when the client goes to the background and back to foreground. Sometimes the order or async work and system overload ends up in an order or execution where the start is not possible due to cancelation of previous sync.

**Comment by LukasKorba** [comment]
Investigation of this issue actually discovered deeper problem in the loop of CompactBlockProcessor.

Let me summarize first what this ticket was about.

### Initial issue
Zashi is attached to the iOS app's lifecycle, specially when app goes to the background and back to the foreground. These 2 state changes control the synchronizer. Stop() is called when app goes to the background and Start(retry: true) is called when app goes back to the foreground.

As a side effect, sometimes Zashi ended up in stopped state like the start is ignored but was able to recover in 10-30s depending on timer kick. 

### Discovered issue
Generally speaking the SDK and CompactBlockProcessor loop are designed to always recover from errors or unexpected states. Basically every 10-30s the next sync run is triggered so any potential errors are replaced with syncing state. Sometimes the communication or networking fails but it's a false positive so next run works. Even the stopped state not being immediately restarted will be resolved in 10-30s by next sync start. We wanted to reduce these 10-30s delays so the sync is not blocked by waiting.

The mechanism that was supposed to immediately start after stop is in place and triggered the code but run loop in CBP responded with another cancelation error, resulting in stopped state, this stopped state is basically a different one, new one.

And this actually uncovered a very serious flaw in the run loop design. The CBP's run loop lives in a concurrent Task, that one is being canceled so any attempt to continue the run will simply ends up again on a cancelation error. 

What must happen is to stop the run loop and start a brand new Task with the different run loop. This is missing in the code. As a result, run loop ends with any error and CBP waits on a timer to be kicked again (here are those 10-30s of waiting). 

### Proposal for solution
There are several things that should be done, some are design fixes and some are new optimized proposals that I came up with during the debugging session.

1. Restart Task when cancelation error - do not try to continue in the same context
2. Remove timer from the SDK, it's no longer needed as we already have a while true CBP loop that is async one and controls everything. Timer basically wraps up some code and won't let us to control the time (see the next point).
3. With removed timer, we will control time in a concurrent context by sleeping the Task. We will get the same behaviour but newly we will be able to control how much time we wait until next sync. That will allow us to do gradual waiting and retry logic. In case of an error, we will wait 1s (because majority of errors are false positive and the run can simply continue asap). If the error persists the time will get to those 10-30s and stay there, resulting in the same behaviour as we have now. So real errors like server down will be handled in the same manner. 
4. see the whole Known issues section.

### Known issues
With a few changes during the debugging session (basically hotfixes and workarounds for 3. mentioned proposals) some errors started to appear. Mostly related to scanning some blocks that are not present on the disk or writing metadata failures, etc. I believe these errors were always there but hidden by "not trying to recover asap" as well as not starting the synchronizer asap. These errors are probably related to some +-1 computation or even wrong order of calls in the state machine actions. Needs to be investigated.

All in all, current SDK is pretty much safe because it wait for the next sync. Any new sync starts with cleaned up states and data so any errors from the past are cleared up. We still want to reduce those 10-30s of unnecessary waiting time but now we know that it can't be simply done without bigger changes in the CBP's run loop.
